pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/PersistentTopicsTest.java

```diff
@@ -312,4 +316,82 @@ public void testGetPartitionedTopicsList() throws KeeperException, InterruptedEx
         Assert.assertEquals(nonPersistentPartitionedTopics.size(), 1);
         Assert.assertEquals(TopicName.get(nonPersistentPartitionedTopics.get(0)).getDomain().value(), TopicDomain.non_persistent.value());
     }
+
+    @Test
+    public void testGrantNonPartitionedTopic() {
+        final String topicName = "non-partitioned-topic";
+        persistentTopics.createNonPartitionedTopic(testTenant, testNamespace, topicName, true);
+        String role = "role";
+        Set<AuthAction> expectActions = new HashSet<>();
+        expectActions.add(AuthAction.produce);
+        persistentTopics.grantPermissionsOnTopic(testTenant, testNamespace, topicName, role, expectActions);
+        Map<String, Set<AuthAction>> permissions = persistentTopics.getPermissionsOnTopic(testTenant, testNamespace, topicName);
+        Assert.assertEquals(permissions.get(role), expectActions);
+    }
+
+    @Test
+    public void testGrantPartitionedTopic() {
+        final String partitionedTopicName = "partitioned-topic";
+        final int numPartitions = 5;
+        LocalZooKeeperCacheService mockLocalZooKeeperCacheService = mock(LocalZooKeeperCacheService.class);
+        ZooKeeperChildrenCache mockZooKeeperChildrenCache = mock(ZooKeeperChildrenCache.class);
+        doReturn(mockLocalZooKeeperCacheService).when(pulsar).getLocalZkCacheService();
+        doReturn(mockZooKeeperChildrenCache).when(mockLocalZooKeeperCacheService).managedLedgerListCache();
```

Comment:
Mock is not required here.

The following code can be used to check:

```
    @Test
    public void testGrantPartitionedTopic() throws Exception {
        final String partitionedTopicName = "partitioned-topic";
        final int numPartitions = 5;
        persistentTopics.createPartitionedTopic(testTenant, testNamespace, partitionedTopicName, numPartitions);

        String role = "role";
        Set<AuthAction> expectActions = new HashSet<>();
        expectActions.add(AuthAction.produce);
        persistentTopics.grantPermissionsOnTopic(testTenant, testNamespace, partitionedTopicName, role, expectActions);
        for (int i=0; i < numPartitions; i++) {
            Map<String, Set<AuthAction>> partitionPermissions = persistentTopics.getPermissionsOnTopic(testTenant,
                    testNamespace, partitionedTopicName + "-partition-" + i);
            Assert.assertEquals(partitionPermissions.get(role), expectActions);
        }
    }
```
The `getPermissionsOnTopic` function does not seem to support searching partitioned topic yet. I think you can consider adding a feature to the `getPermissionsOnTopic` function to support searching partition topic in this pr or by opening a new pr.

If the `getPermissionsOnTopic` function supports finding partitioned topic, you can directly call the `getPermissionsOnTopic` function to get the value of permissions, so this loop is not needed.

Reply:
Hi, I went through the grant and get logic for permissions and It seems current logic support get permissions for partitions of a topic. Pls help me check if my understanding is right.
The permissions will be stored in a map named destination_auth as followed. 
```java
    private void grantPermissions(String topicUri, String role, Set<AuthAction> actions) {
        try {
            ...
            Policies policies = jsonMapper().readValue(content, Policies.class);

            if (!policies.auth_policies.destination_auth.containsKey(topicUri)) {
                policies.auth_policies.destination_auth.put(topicUri, new TreeMap<String, Set<AuthAction>>());
            }
            policies.auth_policies.destination_auth.get(topicUri).put(role, actions);

            // Write the new policies to zookeeper
            globalZk().setData(path(POLICIES, namespaceName.toString()), jsonMapper().writeValueAsBytes(policies),
                    nodeStat.getVersion());
        ...
        }
```
Then get permissions logic for a partition: try to get permissions from  auth_policies.destination_auth directly by topic_name.
```java
protected Map<String, Set<AuthAction>> internalGetPermissionsOnTopic() {
        // This operation should be reading from zookeeper and it should be allowed without having admin privileges
        validateAdminAccessForTenant(namespaceName.getTenant());

        String topicUri = topicName.toString();

        try {
            ...
            // Then add topic level permissions
            if (auth.destination_auth.containsKey(topicUri)) {
                for (Map.Entry<String, Set<AuthAction>> entry : auth.destination_auth.get(topicUri).entrySet()) {
                    String role = entry.getKey();
                    Set<AuthAction> topicPermissions = entry.getValue();

                    if (!permissions.containsKey(role)) {
                        permissions.put(role, topicPermissions);
                    } else {
                        // Do the union between namespace and topic level
                        Set<AuthAction> union = Sets.union(permissions.get(role), topicPermissions);
                        permissions.put(role, union);
                    }
                }
            }

            return permissions;
        } catch (Exception e) {
            ...
        }
    }
```

Reply:
First of all, this logic has no problem with non-partitioned topics.

Let's look at the creation of the partitioned topic first. When we use the following command to create a partitioned topic, 

```
./bin/pulsar-admin topics create-partitioned-topic test-topic -p 2
```

we will eventually generate the following topic in the zookeeper.

```
persistent://public/default/test-topic-partition-0
persistent://public/default/test-topic-partition-1
```

 When we use the `./bin/pulsar-admin topics grant-permissions` command to authorize for a partitioned topic, we are actually authorizing these topics:

```
persistent://public/default/test-topic-partition-0
persistent://public/default/test-topic-partition-1
```

Therefore, at present, when we get permission, we also need to provide the following topic name:

```
persistent://public/default/test-topic-partition-0
persistent://public/default/test-topic-partition-1
```

However, in actual use, users prefer to provide the following name to get all the permissions under the topic.

```
persistent://public/default/test-topic
```

So we hope that if you have time or are interested, you can provide this support in the `internalGetPermissionsOnTopic` function.

```
if the topic is partitioned topic
   Loop to get permissions for each topic
else
   Logic unchanged
```

Reply:
Thanks for the explanation, I understand.
I will open a new pr to support searching partitioned in **getPermissionsOnTopic**

Reply:
you can provide this support in the internalGetPermissionsOnTopic function.
===========================================================
Is it better to create a new api called getPermissionsOnPartitionedTopic?
I understand if I support searching partitioned topic in internalGetPermissionsOnTopic. The output structure of getPermissionsOnTopic will be changed to
Map<String, Map<String, Set<AuthAction>> which will break API compatibility?

Reply:
> > you can provide this support in the internalGetPermissionsOnTopic function.
> 
> Is it better to create a new api called getPermissionsOnPartitionedTopic?
> I understand if I support searching partitioned topic in internalGetPermissionsOnTopic. The output structure of getPermissionsOnTopic will be changed to
> Map<String, Map<String, Set> which will break API compatibility?

@tuteng Can you give me some advice about this?

Reply:
I am not sure whether it is more appropriate to open a new API interface, because the current authorization and cancellation of permissions for both partitioned topic and non-partitioned topic are done under the same rest API. if opening a new rest API to get permissions causes confusion to users, what do you think? @sijie @jiazhai

Reply:
Great. I am free to have a new api. we could following other apis, such as stats/list .